File: templates/skeleton/go/internal/health/health.go
// Package health exposes the liveness and readiness endpoints (AGENTS.md A8,
// docs/specs/SPEC-001-health-endpoints.md).
package health

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
)

// Dependency is one critical dependency the service needs to serve traffic.
type Dependency struct {
	Name  string
	Check func(context.Context) error
}

// Handler answers the health endpoints.
type Handler struct {
	logger       *slog.Logger
	dependencies []Dependency
}

// NewHandler builds a handler over the given dependencies. Register one entry
// per critical dependency declared in section B2 — database, cache, broker.
func NewHandler(logger *slog.Logger, dependencies ...Dependency) *Handler {
	return &Handler{logger: logger, dependencies: dependencies}
}

// Routes registers the health endpoints on the given mux.
func (h *Handler) Routes(mux *http.ServeMux) {
	mux.HandleFunc("GET /health/live", h.live)
	mux.HandleFunc("GET /health/ready", h.ready)
}

// live reports that the process is running. It touches no dependency.
func (h *Handler) live(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, map[string]any{"status": "alive"})
}

// ready reports whether every critical dependency is reachable.
func (h *Handler) ready(w http.ResponseWriter, r *http.Request) {
	failing := make([]string, 0, len(h.dependencies))
	for _, dependency := range h.dependencies {
		if err := dependency.Check(r.Context()); err != nil {
			// The dependency name is safe to expose; its DSN is not (A5).
			h.logger.WarnContext(r.Context(), "dependency unavailable",
				slog.String("dependency", dependency.Name),
				slog.String("error", err.Error()))
			failing = append(failing, dependency.Name)
		}
	}

	if len(failing) > 0 {
		writeJSON(w, http.StatusServiceUnavailable, map[string]any{
			"status":  "unavailable",
			"failing": failing,
		})
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{"status": "ready"})
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

File: templates/skeleton/python-fastapi/src/@@PROJECT_SNAKE@@/routers/health.py
"""Health endpoints (AGENTS.md A8, SPEC-001).

``/health/live``  — the process is running. Touches nothing.
``/health/ready`` — every critical dependency answers. 503 when one does not.
"""

import logging

from fastapi import APIRouter, Response, status

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/health", tags=["health"])


async def _check_dependencies() -> dict[str, bool]:
    """Probe every critical dependency declared in B2.

    Add one entry per dependency (database, cache, broker). Keep each probe
    cheap and bounded: this endpoint must answer in under 100 ms (SPEC-001).
    """
    return {}


@router.get("/live", status_code=status.HTTP_200_OK)
async def live() -> dict[str, str]:
    """Return 200 as long as the process is able to serve."""
    return {"status": "alive"}


@router.get("/ready")
async def ready(response: Response) -> dict[str, object]:
    """Return 200 when every dependency is reachable, 503 otherwise."""
    checks = await _check_dependencies()
    failing = [name for name, healthy in checks.items() if not healthy]

    if failing:
        # The dependency name is safe to expose; its DSN is not (A5).
        logger.warning("readiness check failed", extra={"failing": failing})
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
        return {"status": "unavailable", "failing": failing}

    return {"status": "ready", "checks": checks}
